express-notes-api/server.js

```diff
@@ -0,0 +1,83 @@
+import express from 'express';
+import { readFile, writeFile } from 'node:fs/promises';
+const app = express();
+app.use(express.json());
+
+async function readData() {
+  const file = await readFile('data.json');
+  return await JSON.parse(file);
+}
+const data = await readData();
+async function dataToJson() {
+  const newObj = JSON.stringify(data, null, 2);
+  await writeFile('data.json', newObj);
+}
+
+app.get('/api/notes', (req, res) => {
+  const arr = [];
+  for (const key in data.notes) {
+    arr.push(data.notes[key]);
+  }
+  res.status(200).json(arr);
+});
+
+app.get('/api/notes/:id', (req, res) => {
+  const note = req.params.id;
+  if (Math.sign(note) !== 1) {
+    res.status(400).send({ error: 'id must be a positive integer' });
+  } else if (!data.notes[note]) {
+    res.status(404).send({ error: 'cannot find note with id ' + note });
+  } else {
+    res.status(200).send(data.notes[note]);
+  }
+});
+
+app.post('/api/notes', (req, res) => {
+  if (req.body === undefined) {
+    res.status(400).send({ error: 'content is a required field' });
+  } else if (req === Error) {
+    res.status(500).send({ error: 'An unexpected error occured.' });
+  } else {
+    const newNote = data.nextId;
+    data.notes[newNote] = req.body;
+    data.notes[newNote].id = data.nextId++;
+    dataToJson();
+    res.send(req.body).json(data.notes[newNote]).status(201);
+  }
+});
+
+app.delete('/api/notes/:id', (req, res) => {
+  const note = req.params.id;
+  if (Math.sign(note) !== 1) {
+    res.status(400).send({ error: 'id must be a positive integer' });
+  } else if (!data.notes[note]) {
+    res.status(404).send({ error: 'cannot find note with id ' + note });
+  } else if (req.body === Error) {
+    res.status(500).send({ error: 'An unexpected error occured.' });
+  } else {
+    delete data.notes[note];
+    dataToJson();
+    res.status(204).send(res.body);
+  }
+});
+
+app.put('/api/notes/:id', (req, res) => {
+  const note = req.params.id;
+  if (Math.sign(note) !== 1) {
+    res.status(400).send({ error: 'id must be a positive integer' });
+  } else if (!data.notes[note]) {
+    res.status(404).send({ error: 'cannot find note with id ' + note });
+  } else if (req.body === Error) {
+    res.status(500).send({ error: 'An unexpected error occured.' });
+  } else if (req.body.content === undefined) {
+    res.status(400).send({ error: 'content field is required' });
+  } else {
+    data.notes[note].id = req.params.id;
+    dataToJson();
+    res.status(204).send(req.body);
```

Comment:
Does this work? I don't see you updating the `data` object with the note in the request. Try a `put` followed by a `get` to make sure you get the updated object back.

Reply:
This works because the dataToJson writes the updated data into the file, i updated it for clarity, and im not sure what you mean by a put requests followed by a get request? Would this be using next()?

Reply:
I meant to do that from the command line to test it. Something like:
```
http put :8080/api/notes/20 id=20 content="New values"
http get :8080/api/notes/20
```
Then compare that the `get` returns the updated note with id 20.

I understand your reasoning and why it worked before. The new version makes it more clear that it works. Well done.